workflow/controller/operator.go

```diff
@@ -1685,7 +1685,8 @@ func (woc *wfOperationCtx) executeTemplate(ctx context.Context, nodeName string,
 			// Inject the retryAttempt number
 			localParams[common.LocalVarRetries] = strconv.Itoa(len(retryParentNode.Children))
 
-			processedTmpl, err = common.SubstituteParams(processedTmpl, map[string]string{}, localParams)
+			// Re-process the args to make sure {{retries}} variable is correctly updated
+			processedTmpl, err = common.ProcessArgs(resolvedTmpl, &args, woc.globalParams, localParams, false)
```

Comment:
Should this not be
```suggestion
			// Re-process the args to make sure {{retries}} variable is correctly updated
			processedTmpl, err = common.ProcessArgs(processedTmpl, &args, woc.globalParams, localParams, false)
```
?

Reply:
Hi. Unfortunately, `processedTmpl` won't work here. Its `PodSpecPatch` is already processed and `{{retries}}` is replaced with `0`. The raw `PodSpecPatch` is lost. Only `resolvedTmpl` has the original `PodSpecPatch`.

`processedTmpl.PodSpecPatch`:
```yaml
containers:
  - name: main
    resources:
      limits:
        memory: "64Mi"
```

`resolvedTmpl.PodSpecPatch`:
```yaml
containers:
  - name: main
    resources:
      limits:
        memory: "{{=(sprig.int(retries) + 1) * 64}}Mi"
```

Another option is that we store the original `PodSpecPatch` as a copy, and substitute args with that. This might work, haven't tested tho:

```diff
                localParams[common.LocalVarPodName] = woc.wf.NodeID(nodeName)
        }
 
+       originalPodSpecPatch := resolvedTmpl.PodSpecPatch
+
        // Inputs has been processed with arguments already, so pass empty arguments.
        processedTmpl, err := common.ProcessArgs(resolvedTmpl, &args, woc.globalParams, localParams, false)
        if err != nil {
@@ -1685,8 +1687,10 @@ func (woc *wfOperationCtx) executeTemplate(ctx context.Context, nodeName string,
                        // Inject the retryAttempt number
                        localParams[common.LocalVarRetries] = strconv.Itoa(len(retryParentNode.Children))
 
+                       // Replace PodSPecPatch with the original one to make sure variables are updated
+                       processedTmpl.PodSpecPatch = originalPodSpecPatch
+
                        processedTmpl, err = common.SubstituteParams(processedTmpl, map[string]string{}, localParams)
```

Let me know what you think. Thanks!

Reply:
Ah I see. My worry is if we use `resolvedTmpl` again would we not lose all of the other variables we have already processed? For example:

```yaml
arguments:
  parameters:
    - name: message
      value: "This is retry #{{retries}} on pod {{pod.name}}"
```

Would your code not output: `value: "This is retry #0 on pod {{pod.name}}"`? Correct me if I'm wrong

Reply:
Ah, never mind. I see you're using `ProcessArgs`. Disregard that.

Reply:
The reason why this worked before and does not work now is because before expressions where allowed, the first call to `ProcessArgs` would leave `{{retries}}` as-is because it didn't then exist in the environment. Then, when we used `SubstituteParams` here we could be sure that a `{{retires}}` (if any) was still not resolved.

However, with the expression, I assume sprint is evaluating `sprig.int(retries)` to 0 because it doesn't exist?

Reply:
Seems like even if we tell the original call to ignore any field with `retries` and leave it as-is, we will have to call `ProcessArgs` anyway... we won't be able to get away with `SubstituteParams` anymore

Reply:
I don't think your proposed solution in https://github.com/argoproj/argo-workflows/pull/5389#discussion_r593927074 would work because I don't believe `SubstituteParams` evaluates expressions

Reply:
I'm trying to see if we can avoid two calls to `ProcessArgs` (which are a bit expensive)... but it seems like a chicken-and-egg problem here

Reply:
Any ideas?

Reply:
> I assume sprint is evaluating sprig.int(retries) to 0 because it doesn't exist?

[This is actually the case](https://github.com/argoproj/argo-workflows/issues/5399)

Reply:
Thanks for the feedback!

> I assume sprint is evaluating sprig.int(retries) to 0 because it doesn't exist?

Yes. Looks like `sprig` will resolve `{{retries}}` to 0 in the original call. I used `sprig` because `{{retries}}` is actually a string here. Without `sprig` we can't do things like `{{=retries + 1}}`:

![Screen Shot 2021-03-14 at 11 41 17 PM](https://user-images.githubusercontent.com/1311594/111100825-fd4b0f80-851e-11eb-9afc-ee4308707087.png)

Let me see what I can do. Thanks!

Reply:
> Seems like even if we tell the original call to ignore any field with retries and leave it as-is, we will have to call ProcessArgs anyway... we won't be able to get away with SubstituteParams anymore

Actually I do think `SubstituteParams` evaluates expressions. I was able to pull it off with this fix only:

```diff
diff --git a/util/template/expression_template.go b/util/template/expression_template.go
index 39bd6f01..86840ada 100644
--- a/util/template/expression_template.go
+++ b/util/template/expression_template.go
@@ -4,6 +4,7 @@ import (
        "fmt"
        "io"
        "os"
+       "strings"

        "github.com/antonmedv/expr"
 )
@@ -15,6 +16,12 @@ func init() {
 }

 func expressionReplace(w io.Writer, expression string, env map[string]interface{}, allowUnresolved bool) (int, error) {
+       if strings.Contains(expression, "retries") {
+               _, ok := env["retries"]
+               if !ok && allowUnresolved {
+                       return w.Write([]byte(fmt.Sprintf("{{%s%s}}", kindExpression, expression)))
+               }
+       }
        result, err := expr.Eval(expression, env)
        if (err != nil || result == nil) && allowUnresolved { //  <nil> result is also un-resolved, and any error can be unresolved
                return w.Write([]byte(fmt.Sprintf("{{%s%s}}", kindExpression, expression)))
```

Basically skips resolving `retries` when it's not in the `env`. What do we think?